False positive for "clippy::needless_collect"

### Summary

I'm creating a vector from an iterator using `collect`, since I want to iterate over some iterator multiple times.igina
But the original iterator didn't implement `Clone`, so I had to collect the elements into a vector first, so I could get an iterator which implements `Clone`.

In my case the iterator was an iterator over the lines of a file using `BufReader::lines`.
But I tried to find out the exact cause of the issue and generalized/simplified the issue in the example.

### Lint Name

clippy::needless_collect

### Reproducer

I tried this code:

```rust
fn test(elements: impl Iterator<Item = Option<usize>>) {
    let elements: Vec<_> = elements.into_iter().map(Option::unwrap).collect();
    let elements = elements.into_iter();

    for _element in elements.clone() {
        // do something with element
    }

    for _element in elements {
        // do something with element
    }
}
```

I got this message:

```
warning: avoid using `collect()` when not needed
 --> src/main.rs:6:69
  |
6 |     let elements: Vec<_> = elements.into_iter().map(Option::unwrap).collect();
  |                                                                     ^^^^^^^
7 |     let elements = elements.into_iter();
  |                    -------------------- the iterator could be used here instead
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_collect
  = note: `-W clippy::needless-collect` implied by `-W clippy::nursery`
  = help: to override `-W clippy::nursery` add `#[allow(clippy::needless_collect)]`
help: use the original Iterator instead of collecting it and then producing a new one
  |
6 ~     
7 ~     let elements = elements.into_iter().map(Option::unwrap);
  |

```

If I follow this suggestion, my code looks like this:

```rust
fn test(elements: impl Iterator<Item = Option<usize>>) {
    let mut elements = elements.map(Option::unwrap);

    for _element in elements.clone() {
        // do something with element
    }

    for _element in elements {
        // do something with element
    }
}
```

But following this suggestion caused me to get another error:

```
error[E0599]: the method `clone` exists for struct `Map<impl Iterator<Item = Option<usize>>, fn(Option<usize>) -> usize {Option::<usize>::unwrap}>`, but its trait bounds were not satisfied
  --> src/main.rs:8:30
   |
8  |     for _element in elements.clone() {
   |                              ^^^^^
   |
  ::: /home/p11/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/adapters/map.rs:62:1
   |
62 | pub struct Map<I, F> {
   | -------------------- doesn't satisfy `_: Clone`
   |
   = note: the following trait bounds were not satisfied:
           `impl Iterator<Item = Option<usize>>: std::clone::Clone`
           which is required by `std::iter::Map<impl Iterator<Item = Option<usize>>, fn(std::option::Option<usize>) -> usize {std::option::Option::<usize>::unwrap}>: std::clone::Clone`
help: consider restricting the type parameter to satisfy the trait bound
   |
5  | fn test(elements: impl Iterator<Item = Option<usize>>) where impl Iterator<Item = Option<usize>>: std::clone::Clone {
   |                                                        ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
```

This suggestion also isn't helpful, since I can't turn my iterator into one which implements `Clone`.

But I already found a solution which works while writing this:

```rust
fn test(elements: impl Iterator<Item = Option<usize>>) {
    let mut elements = elements.map(Option::unwrap);

    for _element in elements.by_ref() {
        // do something with element
    }

    for _element in elements {
        // do something with element
    }
}
```

Not sure if this solution always works and doesn't cause any other issues.


### Version

rustc 1.80.1 (3f5fd8dd4 2024-08-06)
binary: rustc
commit-hash: 3f5fd8dd41153bc5fdca9427e9e05be2c767ba23
commit-date: 2024-08-06
host: x86_64-unknown-linux-gnu
release: 1.80.1
LLVM version: 18.1.7

### Additional Labels

_No response_

--- a/clippy_lints/src/methods/needless_collect.rs
+++ b/clippy_lints/src/methods/needless_collect.rs
@@ -546,6 +546,12 @@ impl<'tcx> Visitor<'tcx> for IteratorMethodCheckVisitor<'_, 'tcx> {
             && !is_trait_method(self.cx, expr, sym::Iterator)
         {
             return ControlFlow::Break(());
+        } else if let ExprKind::Assign(place, value, _span) = &expr.kind
+            && value.hir_id == self.hir_id_of_expr
+            && let Some(id) = path_to_local(place)
+        {
+            // our iterator was directly assigned to a variable
+            self.hir_id_of_let_binding = Some(id);
         }
         walk_expr(self, expr)
     }
@@ -561,6 +567,7 @@ impl<'tcx> Visitor<'tcx> for IteratorMethodCheckVisitor<'_, 'tcx> {
         }) = &stmt.kind
             && expr.hir_id == self.hir_id_of_expr
         {
+            // our iterator was directly assigned to a variable
             self.hir_id_of_let_binding = Some(*id);
         }
         walk_stmt(self, stmt)

--- a/clippy_lints/src/methods/needless_collect.rs
+++ b/clippy_lints/src/methods/needless_collect.rs
@@ -1,3 +1,5 @@
+use std::ops::ControlFlow;
+
 use super::NEEDLESS_COLLECT;
 use clippy_utils::diagnostics::{span_lint_and_sugg, span_lint_hir_and_then};
 use clippy_utils::source::{snippet, snippet_with_applicability};
@@ -9,9 +11,9 @@ use clippy_utils::{
 };
 use rustc_data_structures::fx::FxHashMap;
 use rustc_errors::{Applicability, MultiSpan};
-use rustc_hir::intravisit::{Visitor, walk_block, walk_expr};
+use rustc_hir::intravisit::{Visitor, walk_block, walk_expr, walk_stmt};
 use rustc_hir::{
-    BindingMode, Block, Expr, ExprKind, HirId, HirIdSet, LetStmt, Mutability, Node, PatKind, Stmt, StmtKind,
+    BindingMode, Block, Expr, ExprKind, HirId, HirIdSet, LetStmt, Mutability, Node, Pat, PatKind, Stmt, StmtKind,
 };
 use rustc_lint::LateContext;
 use rustc_middle::hir::nested_filter;
@@ -100,6 +102,12 @@ pub(super) fn check<'tcx>(
                 let mut used_count_visitor = UsedCountVisitor { cx, id, count: 0 };
                 walk_block(&mut used_count_visitor, block);
                 if used_count_visitor.count > 1 {
+                    return;
+                }
+
+                if let IterFunctionKind::IntoIter(hir_id) = iter_call.func
+                    && !check_iter_expr_used_only_as_iterator(cx, hir_id, block)
+                {
                     return;
                 }
 
@@ -253,7 +261,7 @@ struct IterFunction {
 impl IterFunction {
     fn get_iter_method(&self, cx: &LateContext<'_>) -> String {
         match &self.func {
-            IterFunctionKind::IntoIter => String::new(),
+            IterFunctionKind::IntoIter(_) => String::new(),
             IterFunctionKind::Len => String::from(".count()"),
             IterFunctionKind::IsEmpty => String::from(".next().is_none()"),
             IterFunctionKind::Contains(span) => {
@@ -268,7 +276,7 @@ impl IterFunction {
     }
     fn get_suggestion_text(&self) -> &'static str {
         match &self.func {
-            IterFunctionKind::IntoIter => {
+            IterFunctionKind::IntoIter(_) => {
                 "use the original Iterator instead of collecting it and then producing a new one"
             },
             IterFunctionKind::Len => {
@@ -284,7 +292,7 @@ impl IterFunction {
     }
 }
 enum IterFunctionKind {
-    IntoIter,
+    IntoIter(HirId),
     Len,
     IsEmpty,
     Contains(Span),
@@ -343,7 +351,7 @@ impl<'tcx> Visitor<'tcx> for IterFunctionVisitor<'_, 'tcx> {
                     }
                     match method_name.ident.name.as_str() {
                         "into_iter" => self.uses.push(Some(IterFunction {
-                            func: IterFunctionKind::IntoIter,
+                            func: IterFunctionKind::IntoIter(expr.hir_id),
                             span: expr.span,
                         })),
                         "len" => self.uses.push(Some(IterFunction {
@@ -519,4 +527,55 @@ fn get_captured_ids(cx: &LateContext<'_>, ty: Ty<'_>) -> HirIdSet {
     get_captured_ids_recursive(cx, ty, &mut set);
 
     set
+}
+
+struct IteratorMethodCheckVisitor<'a, 'tcx> {
+    cx: &'a LateContext<'tcx>,
+    hir_id_of_expr: HirId,
+    hir_id_of_let_binding: Option<HirId>,
+}
+
+impl<'tcx> Visitor<'tcx> for IteratorMethodCheckVisitor<'_, 'tcx> {
+    type Result = ControlFlow<()>;
+    fn visit_expr(&mut self, expr: &'tcx Expr<'tcx>) -> ControlFlow<()> {
+        if let ExprKind::MethodCall(_method_name, recv, _args, _) = &expr.kind
+            && (recv.hir_id == self.hir_id_of_expr
+                || self
+                    .hir_id_of_let_binding
+                    .is_some_and(|hid| path_to_local_id(recv, hid)))
+            && !is_trait_method(self.cx, expr, sym::Iterator)
+        {
+            return ControlFlow::Break(());
+        }
+        walk_expr(self, expr)
+    }
+    fn visit_stmt(&mut self, stmt: &'tcx Stmt<'tcx>) -> ControlFlow<()> {
+        if let StmtKind::Let(LetStmt {
+            init: Some(expr),
+            pat:
+                Pat {
+                    kind: PatKind::Binding(BindingMode::NONE | BindingMode::MUT, id, _, None),
+                    ..
+                },
+            ..
+        }) = &stmt.kind
+            && expr.hir_id == self.hir_id_of_expr
+        {
+            self.hir_id_of_let_binding = Some(*id);
+        }
+        walk_stmt(self, stmt)
+    }
+}
+
+fn check_iter_expr_used_only_as_iterator<'tcx>(
+    cx: &LateContext<'tcx>,
+    hir_id_of_expr: HirId,
+    block: &'tcx Block<'tcx>,
+) -> bool {
+    let mut visitor = IteratorMethodCheckVisitor {
+        cx,
+        hir_id_of_expr,
+        hir_id_of_let_binding: None,
+    };
+    visitor.visit_block(block).is_continue()
 }

--- a/tests/ui/needless_collect.fixed
+++ b/tests/ui/needless_collect.fixed
@@ -87,6 +87,15 @@ fn main() {
     let my_collection: Vec<()> = vec![()].into_iter().map(|()| {}).collect();
     let my_iter = my_collection.into_iter();
     let _sliced = my_iter.as_slice();
+    // Assignment outside of main scope
+    {
+        let x;
+        {
+            let xxx: Vec<()> = vec![()].into_iter().map(|()| {}).collect();
+            x = xxx.into_iter();
+            for i in x.as_slice() {}
+        }
+    }
 }
 
 fn foo(_: impl IntoIterator<Item = usize>) {}

--- a/tests/ui/needless_collect.fixed
+++ b/tests/ui/needless_collect.fixed
@@ -73,6 +73,20 @@ fn main() {
     let mut out = vec![];
     values.iter().cloned().map(|x| out.push(x)).collect::<Vec<_>>();
     let _y = values.iter().cloned().map(|x| out.push(x)).collect::<Vec<_>>(); // this is fine
+
+    // Don't write a warning if we call `clone()` on the iterator
+    // https://github.com/rust-lang/rust-clippy/issues/13430
+    let my_collection: Vec<()> = vec![()].into_iter().map(|()| {}).collect();
+    let _cloned = my_collection.into_iter().clone();
+    let my_collection: Vec<()> = vec![()].into_iter().map(|()| {}).collect();
+    let my_iter = my_collection.into_iter();
+    let _cloned = my_iter.clone();
+    // Same for `as_slice()`, for same reason.
+    let my_collection: Vec<()> = vec![()].into_iter().map(|()| {}).collect();
+    let _sliced = my_collection.into_iter().as_slice();
+    let my_collection: Vec<()> = vec![()].into_iter().map(|()| {}).collect();
+    let my_iter = my_collection.into_iter();
+    let _sliced = my_iter.as_slice();
 }
 
 fn foo(_: impl IntoIterator<Item = usize>) {}

--- a/tests/ui/needless_collect.rs
+++ b/tests/ui/needless_collect.rs
@@ -87,6 +87,15 @@ fn main() {
     let my_collection: Vec<()> = vec![()].into_iter().map(|()| {}).collect();
     let my_iter = my_collection.into_iter();
     let _sliced = my_iter.as_slice();
+    // Assignment outside of main scope
+    {
+        let x;
+        {
+            let xxx: Vec<()> = vec![()].into_iter().map(|()| {}).collect();
+            x = xxx.into_iter();
+            for i in x.as_slice() {}
+        }
+    }
 }
 
 fn foo(_: impl IntoIterator<Item = usize>) {}

--- a/tests/ui/needless_collect.rs
+++ b/tests/ui/needless_collect.rs
@@ -73,6 +73,20 @@ fn main() {
     let mut out = vec![];
     values.iter().cloned().map(|x| out.push(x)).collect::<Vec<_>>();
     let _y = values.iter().cloned().map(|x| out.push(x)).collect::<Vec<_>>(); // this is fine
+
+    // Don't write a warning if we call `clone()` on the iterator
+    // https://github.com/rust-lang/rust-clippy/issues/13430
+    let my_collection: Vec<()> = vec![()].into_iter().map(|()| {}).collect();
+    let _cloned = my_collection.into_iter().clone();
+    let my_collection: Vec<()> = vec![()].into_iter().map(|()| {}).collect();
+    let my_iter = my_collection.into_iter();
+    let _cloned = my_iter.clone();
+    // Same for `as_slice()`, for same reason.
+    let my_collection: Vec<()> = vec![()].into_iter().map(|()| {}).collect();
+    let _sliced = my_collection.into_iter().as_slice();
+    let my_collection: Vec<()> = vec![()].into_iter().map(|()| {}).collect();
+    let my_iter = my_collection.into_iter();
+    let _sliced = my_iter.as_slice();
 }
 
 fn foo(_: impl IntoIterator<Item = usize>) {}

